### ISSUE TITLE
PHP-FPM: Another FPM instance seems to already listen on

### ISSUE THREAD

**Issue by iliaross** [virtualmin/virtualmin-gpl#630]
Jamie, can you think of what could (if still could) cause not to clear old PHP-FPM config? We end-up listening on the same PHP-FPM socket, and PHP-FPM fails to restart e.g.:

<img width="1440" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/72bb46fc-b7c2-4490-8f88-b1130fb823b8">

I tried on Ubuntu 22.04, with the default domain.

Can you please check the code one more time, because I cannot really reproduce it with normal workflow, with latest code applied but it certainly something introduced by Virtualmin.

I know we added a lot of fixes to _master_ to clean old PHP-FPM config .. I just would like to be 100% certain that this issue for 7.8 release will be solved.

**Comment by iliajie** [comment]
To be clear, with the current code, if the system is _already_ in this state it's _not possible_ to fix this issue unless PHP version is changed for the virtual server.

For instance, in my case I had PHP 8.3 set for the virtual server. So, just saving **PHP Options** page would result in the  error from the screenshot above. Although, changing PHP version made things go back to normal. But it may not be straight forward.

**Comment by jcameron** [comment]
Can you explain the exact steps that triggered this bug? Does it happen that when changing the PHP version, the only FPM config is left behind and still listens on the old port?

**Comment by iliajie** [comment]
> Can you explain the exact steps that triggered this bug?

I cannot, unfortunately.

> Does it happen that when changing the PHP version, the only FPM config is left behind and still listens on the old port?

It happens if the configs are broken already (can easily be replicated using screenshot), and if **PHP Options** page is saved. However, if **PHP Options** page saved and PHP version is changed, then it auto-heals itself.

**Comment by jcameron** [comment]
Ok, but how did it get into that broken state?

**Comment by iliajie** [comment]
> Ok, but how did it get into that broken state?

I cannot tell .. I simply don't know!

But, don't you think, that saving **PHP Options** page should fix this problem automatically, without the need of changing PHP version first?

**Comment by jcameron** [comment]
No, I think that would be too complex for it to re-write a bunch of FPM configs in the case of a potential port clash. I'd rather we figure out the root cause of this..

One possible cause is the bug I fixed in https://github.com/virtualmin/virtualmin-gpl/commit/b4152d8162792f14b7141e5c4af91df280febbcc

**Comment by iliajie** [comment]
.. or maybe first catch if PHP-FPM service for the given PHP version cannot be restarted and if not, then try to fix it the same way if we changed PHP version for the domain?

**Comment by iliajie** [comment]
> One possible cause is the bug I fixed in [b4152d8](https://github.com/virtualmin/virtualmin-gpl/commit/b4152d8162792f14b7141e5c4af91df280febbcc)

But this patch is related to Apache?

**Comment by jcameron** [comment]
Oh, so did you only see this on an Nginx system?

**Comment by iliajie** [comment]
> Oh, so did you only see this on an Nginx system?

The initial question isn't about Apache or Nginx, it is about PHP-FPM config file.

**Comment by iliajie** [comment]
> Right, but the code that updates the FPM config is called from the Apache and Nginx modules. So it may matter which webserver you're using..

For me the bug happened on the system running Apache.

**Comment by jcameron** [comment]
Ok ... I'm afraid I haven't been able to figure out what causes this issue, sorry.

**Comment by iliajie** [comment]
The fix (in my view), if FPM server cannot be restarted would first attempt to disable PHP and then re-enable it with the same PHP-FPM execution mode and the same PHP version. If that fails again, then show the error as we do now.

Upon disable, it should clear all configs first and then, yes, upon re-enabling reallocate the port/socket.

**Comment by jcameron** [comment]
Ok I'll think about this and see what I can do ..

**Comment by iliajie** [comment]
Thanks!

**Comment by jcameron** [comment]
I've added code to detect this situation, and adding a warning on the PHP Options page and a button to fix it.

**Comment by iliajie** [comment]
> I've added code to detect this situation, and adding a warning on the PHP Options page and a button to fix it.

Got it, thank you! I see the code to check for the port clash. But my initial problem was related to Unix socket clash, so to say, e.g.:

<img width="807" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/774909a4-8c7f-4b26-8d21-389225c0f0fb">

Also, can we please improve this new message in alert box:

<img width="1178" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/e1d1dfc6-4653-4b01-a21d-0c5851d630d6">

Could you explain what's wrong on, and what we're going to do, and that it will happen on the page save .. instead of saying **Fixing PHP FPM port configuration ..** because I'm pretty sure there will be "issue" reports saying, **"I've waited for an hour and nothing happens .."**

And finally, shouldn't we add fix mode it to Virtualmin CLI as well?

**Comment by jcameron** [comment]
Did you pull all my most recent changes and restart webmin? The error message is more descriptive for me ...

And it should already detect FPM socket file clashes as well.

**Comment by iliajie** [comment]
> The error message is more descriptive for me

Oh, it was a duplicate key. Fixed in this patch https://github.com/virtualmin/virtualmin-gpl/commit/75de92014ff6d403b9adfdbc829e2632a599718f.

> And it should already detect FPM socket file clashes as well.

It doesn't detect FPM socket clash correctly because it could be used by the same domain, as was reported in my original issue!

I have taken a closer look and made a fix for this is issue in this patch [dd29039](https://github.com/virtualmin/virtualmin-gpl/commit/dd290394f3e1d078ef45fdf1f8ce45bf06314aea#diff-aa98c4c01f2f6cde518949f907eb960ca25f3f210a756e7729ad4c503bdd8ddd).

However, the clash can also be caused (theoretically) by the same socket being used in another domain, so to address this issue I added another patch. It will also make sure that the error message will be shown correctly to the server owner if some other domain is using the socket and the given server owner doesn't have access to edit its PHP options https://github.com/virtualmin/virtualmin-gpl/commit/f5011cc634d00dfee0a7d4ce1e0b3dc78ad33c16 and later this https://github.com/virtualmin/virtualmin-gpl/commit/30c94209f5677a8e79704f65cc5b61ccd8e1a31e.

Finally, we still need one more patch, because it's not possible to fix **PHP-FPM socket clash** (original issue) unless PHP version is change, i.e. doing only this:

```
&save_domain_php_mode($d, "cgi");
&save_domain_php_mode($d, "fpm");
```

.. in `save_phpmode.cgi` is not enough, we need to change PHP version too within FPM mode ..

**Comment by iliajie** [comment]
> Finally, we still need one more patch, because it's not possible to fix **PHP-FPM socket clash** (original issue) unless PHP version is change, i.e. doing only this:
> 
> ```
> &save_domain_php_mode($d, "cgi");
> &save_domain_php_mode($d, "fpm");
> ```
> 
> .. in `save_phpmode.cgi` is not enough, we need to change PHP version too within FPM mode ..

No need to change PHP version actually. All we need is to clean stray FPM config. Here is the patch https://github.com/virtualmin/virtualmin-gpl/commit/a761e4c4025d4917f8e5598eeef3fb433c22b630.

**Comment by jcameron** [comment]
Ok let me take a look at your patches...

Why link to the other clashing domain to fix the clash though?

**Comment by iliajie** [comment]
> Why link to the other clashing domain to fix the clash though?

Because in order to make a fix, we need to run it on the right domain.

> Ok let me take a look at your patches...

Thanks!

**Comment by jcameron** [comment]
Ok but if they're both using the same port or socket file, those two pages will just link to each other! There's no way to know necessarily which domain is the "real" owner of a port or socket.

I will make some code changes to handle this better...

**Comment by iliajie** [comment]
> Ok but if they're both using the same port or socket file, those two pages will just link to each other! 

Hmm.. I couldn't reproduce it in my tests.

**Comment by jcameron** [comment]
Please send me PRs for these kinds of changes in future instead of just committing, because I don't think these changes work the way you think they do, and it's tricky to undo multiple patches after they've been committed.

**Comment by iliajie** [comment]
Yeah! Sorry! I realized after the second patch that it had to be a PR. I was going to do a small change and then things got more complicated.

What exactly isn't working the expected way?

**Comment by iliajie** [comment]
I can see your latest commit. I can tell that we now just always delete any socket associated with the domain by default.

Is there anything else?

**Comment by jcameron** [comment]
Yeah, hold on while I work on an improved fix ...

**Comment by iliajie** [comment]
Sure! Thanks!

**Comment by jcameron** [comment]
Ok check out https://github.com/virtualmin/virtualmin-gpl/commit/0ab70c8d1ee08ec913b84d4a1bef4e6612beae1f

**Comment by iliajie** [comment]
I gave it more tests and everything looks really smooth!

However, there is one small problem that could potentially cause PHP-FPM config not being fixed. If there are issues with PHP-FPM service, and Virtualmin configuration is re-checked, Virtualmin stops considering that PHP-FPM version exists .. and as the same code is used when fixing PHP-FPM config, the problemed PHP-FPM version is just skipped.

Can we always include all installed PHP-FPM versions when fixing disregard of errors?

**Comment by jcameron** [comment]
So for the port clash check, we still look at all FPM versions installed even if an error is detected..

**Comment by iliajie** [comment]
> So for the port clash check, we still look at all FPM versions installed even if an error is detected..

Can you show me the function to which you're referring to that checks all FPM versions installed?

**Comment by websmurf** [comment]
Not sure if it's related, but I noticed this since a few weeks (when changing a php version of one of the servers):

The initial settings, as you can see, php 8.1 is the current one:
<img width="1529" alt="1  initial" src="https://github.com/virtualmin/virtualmin-gpl/assets/4027236/daa4073f-b034-4254-a43c-98872e7e0484">

Select PHP 8.2 in stead of PHP 8.1:
<img width="1525" alt="2  select-new-version" src="https://github.com/virtualmin/virtualmin-gpl/assets/4027236/4fb56e69-4920-4e76-b9fb-fb6314e01eec">

It will highlight the save option, once you click it:
<img width="1523" alt="3  restart-log" src="https://github.com/virtualmin/virtualmin-gpl/assets/4027236/e632048c-287b-451a-ac85-61d48419a8da">

As you can see, it restarts the php-5.6 php-fpm service in stead of the one that was selected previously (8.1) resulting in a wrong (and malfunctioning) website.

**Comment by jcameron** [comment]
@websmurf this is a known bug that will be fixed in the next Virtualmin release..

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
